File: bench/ts/ssh-bench/throughputBenchmark.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { Benchmark } from './benchmark';
import {
	SshChannel,
	SshClientCredentials,
	SshClientSession,
	SshSessionConfiguration,
	SshAlgorithms,
} from '@microsoft/dev-tunnels-ssh';
import { SshClient, SshServer } from '@microsoft/dev-tunnels-ssh-tcp';
import { CancellationTokenSource } from 'vscode-jsonrpc';

const MessageCountMeasurement = 'Throughput (msgs/s)';
const ByteCountMeasurement = 'Throughput (MB/s)';

declare type hrtime = [number, number];
const millis = ([s, ns]: hrtime) => s * 1000 + ns / 1000000;

export class ThroughputBenchmark extends Benchmark {
	private readonly port: number;
	private readonly initPromise: Promise<void>;
	private readonly server: SshServer;
	private readonly serverPromise: Promise<void>;
	private readonly client: SshClient;
	private readonly messageData: Buffer;

	// Reuse a single session+channel across all runs (matching Go).
	// Creating a fresh TCP+SSH session per run causes backpressure deadlocks
	// with large encrypted messages where TCP send/receive buffers create
	// a circular dependency between the client's send and the server's
	// window update on a "cold" connection.
	private clientSession: SshClientSession | null = null;
	private channel: SshChannel | null = null;

	public constructor(
		private readonly duration: number,
		messageSize: number,
		private readonly withEncryption: boolean,
	) {
		super(
			`Throughput - ${messageSize} byte messages ${
				withEncryption ? 'with' : 'without'
			} encryption`,
			'session-throughput',
			{ encryption: withEncryption ? 'true' : 'false', size: messageSize.toString() },
		);

		const config = new SshSessionConfiguration(withEncryption);
		this.server = new SshServer(config);
		this.client = new SshClient(config);

		this.messageData = Buffer.alloc(messageSize);
		SshAlgorithms.random.getBytes(this.messageData);

		this.initPromise = SshAlgorithms.publicKey
			.rsaWithSha512!.generateKeyPair()
			.then((serverKey) => {
				this.server.credentials.publicKeys.push(serverKey);
			});

		this.server.onSessionOpened((session) => {
			session.onAuthenticating((e) => {
				e.authenticationPromise = Promise.resolve({});
			});
			session.onChannelOpening((e) => {
				e.channel.onDataReceived((data) => {
					e.channel.adjustWindow(data.length);
				});
				e.channel.onRequest((e) => {
					e.isAuthorized = true;
				});
			});
		});

		this.port = Benchmark.findAvailablePort();
		this.serverPromise = this.server.acceptSessions(this.port, 'localhost');
	}

	private async ensureSession(): Promise<void> {
		if (this.channel) return;

		await this.initPromise;

		this.clientSession = await this.client.openSession('localhost', this.port);

		if (this.withEncryption) {
			this.clientSession.onAuthenticating((e) => {
				e.authenticationPromise = Promise.resolve({});
			});

			const credentials: SshClientCredentials = { username: 'benchmark', password: 'benchmark' };
			await this.clientSession.authenticate(credentials);
		}

		this.channel = await this.clientSession.openChannel();
	}

	public async run(): Promise<void> {
		await this.ensureSession();

		// Safety timeout: 4x the benchmark duration catches true deadlocks
		// caused by TCP buffer circular dependencies with large encrypted
		// messages, without affecting normal runs.
		const cancellationSource = new CancellationTokenSource();
		const timeoutHandle = setTimeout(
			() => cancellationSource.cancel(), this.duration * 4);

		let deadlocked = false;
		const startTime: hrtime = process.hrtime();
		let elapsed = 0;

		let messageCount = 0;
		try {
			while (elapsed < this.duration) {
				await this.channel!.send(this.messageData, cancellationSource.token);
				messageCount++;
				elapsed = millis(process.hrtime(startTime));
			}
		} catch {
			deadlocked = true;
		}

		clearTimeout(timeoutHandle);
		cancellationSource.dispose();

		// If the session deadlocked, reset it so the next run gets a fresh one.
		if (deadlocked) {
			this.clientSession?.dispose();
			this.clientSession = null;
			this.channel = null;
		}

		if (messageCount === 0) return;

		elapsed = millis(process.hrtime(startTime));
		const elapsedSeconds = elapsed / 1000;
		const messagesPerSecond = messageCount / elapsedSeconds;
		const bytesPerSecond = (messageCount * this.messageData.length) / elapsedSeconds;
		const megabytesPerSecond = bytesPerSecond / (1024 * 1024);
		this.addMeasurement(MessageCountMeasurement, messagesPerSecond);
		this.addMeasurement(ByteCountMeasurement, megabytesPerSecond);
	}

	public async verify(): Promise<void> {
		await this.initPromise;

		const messageCount = 5;
		let receivedCount = 0;
		const receivedPromise = new Promise<number>((resolve) => {
			const sessionReg = this.server.onSessionOpened((session) => {
				sessionReg.dispose();
				session.onChannelOpening((e) => {
					e.channel.onDataReceived((data) => {
						e.channel.adjustWindow(data.length);
						receivedCount++;
						if (receivedCount >= messageCount) {
							resolve(receivedCount);
						}
					});
				});
			});
		});

		const clientSession = await this.client.openSession('localhost', this.port);

		if (this.withEncryption) {
			clientSession.onAuthenticating((e) => {
				e.authenticationPromise = Promise.resolve({});
			});

			const credentials: SshClientCredentials = { username: 'benchmark', password: 'benchmark' };
			await clientSession.authenticate(credentials);
		}

		const channel = await clientSession.openChannel();

		for (let i = 0; i < messageCount; i++) {
			await channel.send(this.messageData);
		}

		// Wait for server to receive all messages (with timeout)
		const received = await Promise.race([
			receivedPromise,
			new Promise<number>((_, reject) =>
				setTimeout(() => reject(new Error('Timeout waiting for messages')), 5000),
			),
		]);

		if (received !== messageCount) {
			throw new Error(`Expected ${messageCount} messages, received ${received}`);
		}
	}

	public async dispose(): Promise<void> {
		this.clientSession?.dispose();
		this.server.dispose();
		this.client.dispose();
		await this.serverPromise;
	}
}

File: bench/cs/Ssh.Benchmark/ThroughputBenchmark.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net;
using System.Security.Claims;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.DevTunnels.Ssh.Events;
using Microsoft.DevTunnels.Ssh.Messages;
using Microsoft.DevTunnels.Ssh.Tcp;

namespace Microsoft.DevTunnels.Ssh.Benchmark;

#if NETSTANDARD2_0 || NET4
using ValueTask = System.Threading.Tasks.Task;
#endif

class ThroughputBenchmark : Benchmark
{
	private const string MessageCountMeasurement = "Throughput (msgs/s)";
	private const string ByteCountMeasurement = "Throughput (MB/s)";
	private const string BytesAllocatedMeasurement = "Bytes allocated (KB/msg)";
	private const string BytesCopiedMeasurement = "Bytes copied (KB/msg)";

	private readonly TimeSpan duration;
	private readonly SshServer server;
	private readonly Task serverTask;
	private readonly SshClient client;
	private readonly byte[] messageData;
	private readonly bool withEncryption;

	// Reuse a single session+channel across all runs (matching Go).
	// Creating a fresh TCP+SSH session per run causes backpressure deadlocks
	// with large encrypted messages where TCP send/receive buffers create
	// a circular dependency between the client's send and the server's
	// window update on a "cold" connection.
	private SshClientSession clientSession;
	private SshChannel channel;

	public ThroughputBenchmark(TimeSpan duration, int messageSize, bool withEncryption)
		: base(
			$"Throughput - {messageSize} byte messages {(withEncryption ? "with" : "without")} encryption",
			"session-throughput",
			new Dictionary<string, string>
			{
				{ "encryption", withEncryption ? "true" : "false" },
				{ "size", messageSize.ToString() },
			})
	{
		HigherIsBetter[BytesAllocatedMeasurement] = false;
		HigherIsBetter[BytesCopiedMeasurement] = false;

		var trace = new TraceSource(nameof(SessionSetupBenchmark));

		var config = withEncryption
			? SshSessionConfiguration.Default : SshSessionConfiguration.NoSecurity;
		this.server = new SshServer(config, trace);
		this.client = new SshClient(config, trace);

		this.messageData = new byte[messageSize];
		new Random(1).NextBytes(this.messageData);
		this.withEncryption = withEncryption;
		this.duration = duration;

		var serverKey = SshAlgorithms.PublicKey.RsaWithSha512.GenerateKeyPair();
		this.server.Credentials = new[] { serverKey };

		this.server.SessionAuthenticating += OnServerSessionAuthenticating;
		this.server.ChannelRequest += OnServerChannelRequest;

		// Set up server-side data drain for all sessions.
		this.server.SessionOpened += (_, serverSession) =>
		{
			serverSession.ChannelOpening += (__, e) =>
			{
				e.Channel.DataReceived += (___, data) =>
				{
					e.Channel.AdjustWindow((uint)data.Count);
				};
			};
		};

		this.serverTask = this.server.AcceptSessionsAsync(Benchmark.ServerPort, IPAddress.Loopback);
	}

	private async Task EnsureSessionAsync()
	{
		if (this.channel != null) return;

		this.clientSession = await client.OpenSessionAsync(
			IPAddress.Loopback.ToString(), ServerPort);

		if (this.withEncryption)
		{
			this.clientSession.Authenticating += OnClientSessionAuthenticating;
			await this.clientSession.AuthenticateAsync(("benchmark", "benchmark"));
		}

		this.channel = await this.clientSession.OpenChannelAsync();
	}

	protected override async Task RunAsync(Stopwatch stopwatch)
	{
#if DEBUG
		Buffer.Allocations.Clear();
		Buffer.Copies.Clear();
#endif

		await EnsureSessionAsync();

		// Safety timeout: 4x the benchmark duration catches true deadlocks
		// caused by TCP buffer circular dependencies with large encrypted
		// messages, without affecting normal runs.
		using var cancellationSource = new CancellationTokenSource(
			TimeSpan.FromSeconds(this.duration.TotalSeconds * 4));

		bool deadlocked = false;
		int messageCount = 0;
		stopwatch.Restart();
		while (stopwatch.Elapsed < this.duration)
		{
			try
			{
				await this.channel.SendAsync(this.messageData, cancellationSource.Token);
			}
			catch (OperationCanceledException)
			{
				deadlocked = true;
				break;
			}
			messageCount++;
		}

		stopwatch.Stop();

		// If the session deadlocked, reset it so the next run gets a fresh one.
		if (deadlocked)
		{
			this.clientSession?.Dispose();
			this.clientSession = null;
			this.channel = null;
		}

		if (messageCount == 0) return;

		var elapsedSeconds = stopwatch.Elapsed.TotalSeconds;
		double messagesPerSecond = messageCount / elapsedSeconds;
		double bytesPerSecond = (messageCount * this.messageData.Length) / elapsedSeconds;
		double megabytesPerSecond = bytesPerSecond / (1024 * 1024);
		AddMeasurement(MessageCountMeasurement, messagesPerSecond);
		AddMeasurement(ByteCountMeasurement, megabytesPerSecond);

#if DEBUG
		lock (Buffer.Allocations)
		{
			long bytesAllocated = Buffer.Allocations.Values
				.Select((list) => list.Select(count => (long)count).Sum()).Sum();
			double kilobytesPerMessage = (double)bytesAllocated / 1024 / messageCount;
			AddMeasurement(BytesAllocatedMeasurement, kilobytesPerMessage);
		}

		lock (Buffer.Copies)
		{
			long bytesCopied = Buffer.Copies.Values
				.Select((list) => list.Select(count => (long)count).Sum()).Sum();
			double kilobytesPerMessage = (double)bytesCopied / 1024 / messageCount;
			AddMeasurement(BytesCopiedMeasurement, kilobytesPerMessage);
		}
#endif
	}

	private void OnClientSessionAuthenticating(object sender, SshAuthenticatingEventArgs e)
	{
		e.AuthenticationTask = Task.FromResult(new ClaimsPrincipal());
	}

	private void OnServerSessionAuthenticating(object sender, SshAuthenticatingEventArgs e)
	{
		e.AuthenticationTask = Task.FromResult(new ClaimsPrincipal());
	}

	private void OnServerChannelRequest(object sender, SshRequestEventArgs<ChannelRequestMessage> e)
	{
		e.IsAuthorized = true;
	}

	public override async Task VerifyAsync()
	{
		// Send a known amount of data and verify the server receives the correct byte count.
		int totalReceived = 0;
		var receiveDone = new TaskCompletionSource<int>();

		var expectedSize = this.messageData.Length;

		EventHandler<SshServerSession> sessionOpenedHandler = null;
		sessionOpenedHandler = (_, serverSession) =>
		{
			this.server.SessionOpened -= sessionOpenedHandler;
			serverSession.ChannelOpening += (__, e) =>
			{
				e.Channel.DataReceived += (___, data) =>
				{
					e.Channel.AdjustWindow((uint)data.Count);
					var newTotal = Interlocked.Add(ref totalReceived, data.Count);
					if (newTotal >= expectedSize)
						receiveDone.TrySetResult(newTotal);
				};
			};
		};
		this.server.SessionOpened += sessionOpenedHandler;

		using var clientSession = await client.OpenSessionAsync(
			IPAddress.Loopback.ToString(), ServerPort);

		if (this.withEncryption)
		{
			clientSession.Authenticating += OnClientSessionAuthenticating;
			await clientSession.AuthenticateAsync(("benchmark", "benchmark"));
		}

		var channel = await clientSession.OpenChannelAsync();

		await channel.SendAsync(this.messageData, CancellationToken.None);

		var completed = await Task.WhenAny(receiveDone.Task, Task.Delay(5000));
		if (completed != receiveDone.Task)
			throw new Exception($"Timed out: received {totalReceived} of {expectedSize} bytes");

		var received = await receiveDone.Task;
		if (received != expectedSize)
			throw new Exception($"Byte count mismatch: expected {expectedSize}, got {received}");
	}

	public override async ValueTask DisposeAsync()
	{
		this.clientSession?.Dispose();
		this.server.Dispose();
		this.client.Dispose();
		await this.serverTask;
	}
}
